### PR TITLE
feat: new vacuum implementation

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. new storage file name format, add `table_version` in file name,new file format: `{uuid}_{table_seq}_{format_version}`
2. use uuid V7 instead of uuid V4 in snapshot file name.
3. compatable test
4. new vacuum implmentation.

- Closes https://github.com/datafuselabs/databend/issues/12671

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13468)

### CODE REVIEW

**Comment by lichuang** [comment]
askbend:summary

**Comment by BohuTANG** [comment]
askbend:summary

File: src/query/storages/common/table_meta/src/meta/v5/snapshot.rs [hunk 1]
```diff
@@ -28,18 +27,24 @@ use crate::meta::format::encode;
...
 use crate::meta::SnapshotId;
+use crate::meta::Statistics;
 use crate::meta::Versioned;
 
+pub type TableVersion = u64;
+// information to generate snapshot file name
+pub type SnapshotLocationInfo = Option<(SnapshotId, FormatVersion, Option<TableVersion>)>;
+
+/// The structure of the segment is the same as that of v2, but the serialization and deserialization methods are different
 #[derive(Serialize, Deserialize, Clone, Debug)]
```
Comment (zhyass) [inline src/query/storages/common/table_meta/src/meta/v5/snapshot.rs:48]: Maybe we can use the new Statistics https://github.com/datafuselabs/databend/blob/v1.2.184-nightly/src/query/storages/common/table_meta/src/meta/v5/statistics.rs here

**Comment by dantengsky** [comment]
I must have missed something, please correct me if I misunderstood this PR:

- at the beginning of vacuum, there is no `set_table_lvt` action
- during table committing,  lvt is not checked
- if the commit is retried, the snapshot location generated may embed an incorrect table version
   during the retry of table commit, the table will be refreshed (but the table version embedded is supposed to be the table version that the transaction starts with)

File: src/query/storages/fuse/src/fuse_table.rs [hunk 2]
```diff
@@ -286,19 +291,28 @@ impl FuseTable {
         }
     }
 
+    #[minitrace::trace]
+    #[async_backtrace::framed]
```
Comment (zhyass) [inline src/query/storages/fuse/src/fuse_table.rs:295]: Maybe we can use `SnapshotsIO::read_snapshot` directly?
https://github.com/datafuselabs/databend/blob/60004ed525cb2e122e0b5b07810c853ec5b0b617/src/query/storages/fuse/src/io/snapshots.rs#L78-L92

File: src/query/ee/src/storages/fuse/operations/vacuum_table.rs [hunk 3]
```diff
@@ -13,329 +13,510 @@
...
+            //  1. timestamp < retention_time
+            //  2. it is not the current snapshot
+
+            // check if last_snapshot_location'prev match snapshot_file
+            // if not it means that last snapshot or snapshot_files_with_time[i]
+            // has not been committed success
+            // in both cases, return None
+            let last_snapshot = match self
+                .fuse_table
+                .read_table_snapshot_by_location(last_snapshot_location.to_string())
```
Comment (zhyass) [inline src/query/ee/src/storages/fuse/operations/vacuum_table.rs:164]: Maybe we can use `SnapshotsIO::read_snapshot` directly?
https://github.com/datafuselabs/databend/blob/60004ed525cb2e122e0b5b07810c853ec5b0b617/src/query/storages/fuse/src/io/snapshots.rs#L78-L92

File: src/query/ee/src/storages/fuse/operations/vacuum_table.rs [hunk 4]
```diff
@@ -13,329 +13,510 @@
...
-
-    // 2. Get purge orphan segment files.
-    let segment_locations_to_be_purged =
-        get_orphan_files_to_be_purged(fuse_table, referenced_files.segments, retention_time)
+        // purge unreference files
+        let root_gc_snapshot_reference_files = &context.root_gc_snapshot_reference_files;
+        let reference_file_set = vec![
+            &root_gc_snapshot_reference_files.blocks_index,
+            &root_gc_snapshot_reference_files.blocks,
+            &root_gc_snapshot_reference_files.segments,
```
Comment (zhyass) [inline src/query/ee/src/storages/fuse/operations/vacuum_table.rs:394]: Maybe the `table_statistics` and `agg_index` files also need to be purged?

File: src/query/ee/src/storages/fuse/operations/vacuum_table.rs [hunk 5]
```diff
@@ -13,329 +13,510 @@
...
+        batch_purge_files: &[String],
+        purge_files: &mut Option<&mut Vec<String>>,
+    ) -> Result<bool> {
+        if let Some(purge_files) = purge_files {
+            purge_files.extend(batch_purge_files.to_vec());
+            Ok(purge_files.len() >= DRY_RUN_LIMIT)
+        } else {
+            let files: HashSet<String> = batch_purge_files.iter().cloned().collect();
+            self.fuse_table
+                .try_purge_location_files(self.ctx.clone(), files)
```
Comment (zhyass) [inline src/query/ee/src/storages/fuse/operations/vacuum_table.rs:273]: `CompactSegmentInfo`, `BloomIndexMeta` alse need evict cache.

**Comment by drmingdrmer** [comment]
Yes but I do not believe there are so many lines of code changed.

Looks like this PR needs a rebase or merge 🤔

**Comment by lichuang** [comment]
This pr has not been allow to merge for over half a year.

**Comment by BohuTANG** [comment]
Thanks for the PR!
However, the vacuum design is a critical part for us, and testing prior to merging, given its impact on user data. In these situations, taking more time is always better than rushing.

Recommended to initiate an RFC (Request for Comments) followed by a discussion to ensure enougth safety.

**Pull request closed without merging**